fireshape/control.py

```diff
@@ -746,7 +1444,10 @@ def clone(self):
 
     def dot(self, v):
         """Inner product between self and v."""
-        return self.inner_product.eval(self, v)
+        if self.inner_product:
+            return self.inner_product.eval(self, v)
+        else:
+            return self.vec_ro().dot(v.vec_ro())
```

Comment:
this is only justified for wavelets, and since it performed worse, I suggest just reverting to
`return self.inner_product.eval(self, v)`